dvc/cache/base.py

```diff
@@ -101,24 +101,26 @@ def changed(self, path_info, hash_info):
         Args:
             path_info: dict with path information.
             hash: expected hash value for this data.
+            filter_info: an optional argument to target a specific file.
 
         Returns:
             bool: True if data has changed, False otherwise.
         """
 
+        path = filter_info or path_info
         logger.trace(
-            "checking if '%s'('%s') has changed.", path_info, hash_info
+            "checking if '%s' ('%s') has changed.", path_info, hash_info
         )
 
-        if not self.tree.exists(path_info):
-            logger.debug("'%s' doesn't exist.", path_info)
+        if not self.tree.exists(path):
+            logger.debug("'%s' doesn't exist.", path)
             return True
 
         if not hash_info:
             logger.debug("hash value for '%s' is missing.", path_info)
             return True
 
-        if self.changed_cache(hash_info):
+        if self.changed_cache(hash_info, filter_info=filter_info):
             logger.debug(
                 "cache for '%s'('%s') has changed.", path_info, hash_info
             )
```

Comment:
In the logic below we compare `get_hash(path_info)` and `hash_info`, which means that if the directory you are trying to checkout a subdir/subfile of has some unrelated subdir/subfile changed, this method will think that the subdir/subfile is changed and will checkout it again. That's not right.

Reply:
From what I understand, we only have access to the hash_info that already presented in the dvcfile which is usually the base directory. So, this behavior seems correct yet inefficient but I failed to find any better alternative with the current dvcfile format. Do you have any hints for me regarding this?

Reply:
@isidentical Take a look at `set_dir_info` helper. It tries to load `hash_info.dir_info` if the full info for `hash_info` is present in the cache. We could use that too here (or maybe `get_dir_cache` directly) to get a more complete info, then filter it out and use it for a correct comparison. If dir cache is not available - we could indeed stick to the current behavior as a fallback.